### PR TITLE
Implement Keyboard Selection

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Implements the following keyboard selection non-configurable key bindings:
- shift+arrow --> move endpoint by character
- ctrl+shift+left/right --> move endpoint by word
- shift+home/end --> move to beginning/end of line
- ctrl+shift+home/end --> move to beginning/end of buffer

This was purposefully done in the ControlCore layer to make keyboard selection an innate part of how the terminal functions (aka a shared component across terminal consumers).

## References
#715 - Keyboard Selection
#2840 - Spec

## Detailed Description of the Pull Request / Additional comments

The most relevant section is `TerminalSelection.cpp`, where we define how each movement operates. It's basically a giant embedded switch-case statement. We leverage a lot of the work done in a11y to perform the movements.

## Validation Steps Performed
- General cases:
   - test all of the key bindings added
- Corner cases:
   - `char`: wide glyph support
   - `word`: move towards, away, and across the selection pivot
   - automatically scroll viewport
   - ESC (and other key combos) are still clearing the selection properly

### CODE REVIEW

**Comment by carlos-zamora** [comment]
## Discussion
- settings names:
   - "mode" --> "expansionMode"? More descriptive but might not be necessary
   - "updateSelection" --> "moveSelectionAnchor"? Sounds too descriptive imo
   - "view" --> "viewport"? Unsure if "view" is clear, but "viewport" is too in-the-know
   - "buffer" --> ???
   - "cell" --> "char"? We support wide glyphs, so this might be more correct?
- behavior:
   - word selection currently operates like mouse selection (select by runs of same delimiter class, not "words" [i.e. "abc   " is "abc" and "   "]). Text editors behave differently by (un)selecting the whitespace next to the word. Should we consider this a backlog follow-up item? Do we want selection to work this way?
   - buffer selection moves to the end of the full buffer. Something we can't see and that isn't very helpful. Should we set the "end of the buffer" to something more useful/reasonable? Like the bottom of the mutable viewport?
- known issues (unrelated to this PR)
   - rendering issue:
      - select up by viewport; select one cell down; BUG: top line still selected
      - NOTE: select up one cell and down again and it fixes itself

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I guess I'm blocking only for the `GenerateName` thing? And update the docs? But I suppose the second can wait for the naming discussions to be finalized. Code looks great though.

Already looking forward to 1.12!

**Comment by zadjii-msft** [comment]
> * settings names:
>   * "cell" --> "char"? We support wide glyphs, so this might be more correct?

this is the only change I feel is necessary. `char` is better. The rest are fine

> * behavior:
>   * ...Text editors behave differently by (un)selecting the whitespace next to the word. Should we consider this a backlog follow-up item? Do we want selection to work this way?

please no god no. IMO, _bad_ text editors behave this way. Code editors generally don't. 

>   * buffer selection moves to the end of the full buffer. Something we can't see and that isn't very helpful. Should we set the "end of the buffer" to something more useful/reasonable? Like the bottom of the mutable viewport?

Yea, the mutable bottom is more reasonable. The "scroll forward" doesn't exist for all intents and purposes.

**Comment by zadjii-msft** [comment]
_shh bot_

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp [hunk 1]
```diff
@@ -662,4 +663,9 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
                         Id())
         };
     }
+
+    winrt::hstring UpdateSelectionArgs::GenerateName() const
+    {
+        return L"";
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:669]: Yeah, this is a good discussion to have, actually. I think we should hide it from the command palette just because VS Code doesn't have it in its command palette. They have "Expand Selection" and "Shrink Selection" which is a bizarre feature (you should check it out if you're curious).

Honestly, I was even debating removing it as a customizable action in the first place because (like the "copy" action), this can _only_ be used if a selection is active. So we're be claiming _a lot_ of key chords here.

File: src/cascadia/TerminalSettingsModel/defaults.json [hunk 2]
```diff
@@ -358,6 +358,20 @@
...
+        { "command": {"action": "updateSelection", "direction": "down", "mode": "cell" }, "keys": "shift+down" },
+        { "command": {"action": "updateSelection", "direction": "left", "mode": "word" }, "keys": "ctrl+shift+left" },
+        { "command": {"action": "updateSelection", "direction": "right", "mode": "word" }, "keys": "ctrl+shift+right" },
+        { "command": {"action": "updateSelection", "direction": "left", "mode": "view" }, "keys": "shift+home" },
+        { "command": {"action": "updateSelection", "direction": "right", "mode": "view" }, "keys": "shift+end" },
+        { "command": {"action": "updateSelection", "direction": "up", "mode": "view" }, "keys": "shift+pgup" },
+        { "command": {"action": "updateSelection", "direction": "down", "mode": "view" }, "keys": "shift+pgdn" },
+        { "command": {"action": "updateSelection", "direction": "up", "mode": "buffer" }, "keys": "ctrl+shift+home" },
+        { "command": {"action": "updateSelection", "direction": "down", "mode": "buffer" }, "keys": "ctrl+shift+end" },
+
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/defaults.json:374]: Yeah, that can be a separate PR too

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:669]: > I think we should hide it from the command palette

Fine with me. Seems weird to have in the command palette. 

Hard question though: how will it look in the SUI then?

[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:669]: So, there's actually a nasty bug in SUI that this'll make more apparent: if the key binding has no name, it will appear as a nameless key binding in SUI. 😢

Opening a new issue now.

[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:669]: https://github.com/microsoft/terminal/issues/10981

**Comment by carlos-zamora** [comment]
> > buffer selection moves to the end of the full buffer. Something we can't see and that isn't very helpful. Should we set the "end of the buffer" to something more useful/reasonable? Like the bottom of the mutable viewport?

> Yea, the mutable bottom is more reasonable. The "scroll forward" doesn't exist for all intents and purposes.

Updated `_MoveByViewport` and `_MoveByBuffer` to clamp to the mutable bottom. It feels really nice!

**Comment by carlos-zamora** [comment]
It looks like Terminal.hpp can't find `SelectionExpansion` from ICoreSettings.idl when in CI, but it builds just fine locally. Any ideas?

**Comment by zadjii-msft** [comment]
That's so bizarre that this works locally, but not in CI. My guess is related to this bit at the top of Terminal.hpp:

```c++
// You have to forward decl the ICoreSettings here, instead of including the header.
// If you include the header, there will be compilation errors with other
//      headers that include Terminal.hpp
namespace winrt::Microsoft::Terminal::Core
{
    struct ICoreSettings;
    struct ICoreAppearance;
}
```


OH I know why, it's because you're not building `PublicTerminalCore` locally. That's the one that's failing 😋

**Comment by zadjii-msft** [comment]
ad990085 might work to fix this. It certainly did locally. Take it or leave it :P

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 3]
```diff
@@ -35,6 +35,23 @@ namespace winrt::Microsoft::Terminal::Core
 namespace Microsoft::Terminal::Core
 {
     class Terminal;
+
+    enum class InternalSelectionDirection
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/Terminal.hpp:39]: we decided to go with an internal copy of the winrt enum instead of just [including `Microsoft.Terminal.Core` in the WPF Control](https://github.com/microsoft/terminal/commit/ad9900857aa44b3ccbde148cc5ecb53d1a2e5b18) - any reason why?

Comment (DHowett) [reply src/cascadia/TerminalCore/Terminal.hpp:39]: Notes left on commit as to why ;)

Comment (zadjii-msft) [reply src/cascadia/TerminalCore/Terminal.hpp:39]: okay fine yea that felt dirty to write, but I figured the PublicTerminalCore was a. already dirty and b. long overdue for a big refactor that would get rid of the hack anyways

File: src/cascadia/PublicTerminalCore/HwndTerminal.cpp [hunk 4]
```diff
@@ -549,11 +549,11 @@ try
 
     if (multiClickMapper == 3)
     {
-        _terminal->MultiClickSelection(cursorPosition / fontSize, ::Terminal::SelectionExpansionMode::Line);
+        _terminal->MultiClickSelection(cursorPosition / fontSize, InternalSelectionExpansion::Line);
```
Comment (zadjii-msft) [inline src/cascadia/PublicTerminalCore/HwndTerminal.cpp:552]: I'm leave a note behind here saying that once `PublicTerminalCore` is moved to directly consume winrt, we can get rid of `InternalSelectionExpansion` entirely, and just use `winrt::Microsoft::Terminal::Core::SelectionExpansion`

Comment (zadjii-msft) [reply src/cascadia/PublicTerminalCore/HwndTerminal.cpp:552]: #6106 or #6999

**Review by zadjii-msft — DISMISSED** [review]
I hate the duplicated enums, but I understand the world we live in

File: src/cascadia/TerminalCore/Terminal.cpp [hunk 5]
```diff
@@ -862,7 +862,7 @@ WORD Terminal::_TakeVirtualKeyFromLastKeyEvent(const WORD scanCode) noexcept
     return std::unique_lock{ _readWriteLock };
 }
 
-Viewport Terminal::_GetMutableViewport() const noexcept
+Microsoft::Console::Types::Viewport Terminal::_GetMutableViewport() const noexcept
```
Comment (carlos-zamora) [inline src/cascadia/TerminalCore/Terminal.cpp:865]: This is unnecessary

**Comment by carlos-zamora** [comment]
Need to apply changes from [here](https://github.com/microsoft/terminal/pull/2840#issuecomment-921297235). Switching to draft in the meantime.

**Comment by carlos-zamora** [comment]
Assigning @DHowett because he has some very passionate thoughts on keyboard selection haha

**Review by zadjii-msft — APPROVED** [review]
:shipit:

File: src/buffer/out/textBuffer.cpp [hunk 6]
```diff
@@ -1430,12 +1430,13 @@ const til::point TextBuffer::GetGlyphStart(const til::point pos, std::optional<t
...
 // Method Description:
-// - Update pos to be the end of the current glyph/character. This is used for accessibility
+// - Update pos to be the end of the current glyph/character.
 // Arguments:
 // - pos - a COORD on the word you are currently on
+// - accessibilityMode - this is being used for accessibility; make the end exclusive.
 // Return Value:
 // - pos - The COORD for the last cell of the current glyph (exclusive)
-const til::point TextBuffer::GetGlyphEnd(const til::point pos, std::optional<til::point> limitOptional) const
+const til::point TextBuffer::GetGlyphEnd(const til::point pos, bool accessibilityMode, std::optional<til::point> limitOptional) const
```
Comment (lhecker) [inline src/buffer/out/textBuffer.cpp:1439]: The amount of arguments our `TextBuffer` member functions have are a bit of a horror story ngl.
Why does "accessibility mode" mean "become exclusive" here? Don't you mean "bool exclusivePosition"?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 7]
```diff
@@ -33,6 +33,74 @@ constexpr const auto TsfRedrawInterval = std::chrono::milliseconds(100);
...
+    case VK_UP:
+    case VK_PRIOR:
+        return Terminal::SelectionDirection::Up;
+    case VK_DOWN:
+    case VK_NEXT:
+        return Terminal::SelectionDirection::Down;
+    default:
+        return std::nullopt;
+    }
+}
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:55]: This abstraction is only called by one other function. If you inline it like so it'll be immediately obvious which value will be returned for any given input and it saves you a third of the code:
```cpp
using UpdateSelectionParams = std::optional<std::pair<Terminal::SelectionDirection, Terminal::SelectionExpansion>>;

static constexpr UpdateSelectionParams ConvertKeyEventToUpdateSelectionParams(const ControlKeyStates mods, const WORD vkey)
{
    if (mods.IsShiftPressed() && !mods.IsAltPressed())
    {
        if (mods.IsCtrlPressed())
        {
            switch (vkey)
            {
            case VK_LEFT:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Left, Terminal::SelectionExpansion::Word };
            case VK_RIGHT:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Right, Terminal::SelectionExpansion::Word };
            case VK_HOME:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Left, Terminal::SelectionExpansion::Buffer };
            case VK_END:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Right, Terminal::SelectionExpansion::Buffer };
            default:
                [[fallthrough]];
            }
        }
        else
        {
            switch (vkey)
            {
            case VK_HOME:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Left, Terminal::SelectionExpansion::Viewport };
            case VK_END:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Right, Terminal::SelectionExpansion::Viewport };
            case VK_PRIOR:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Up, Terminal::SelectionExpansion::Viewport };
            case VK_NEXT:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Down, Terminal::SelectionExpansion::Viewport };
            case VK_LEFT:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Left, Terminal::SelectionExpansion::Char };
            case VK_RIGHT:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Right, Terminal::SelectionExpansion::Char };
            case VK_UP:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Up, Terminal::SelectionExpansion::Char };
            case VK_DOWN:
                return UpdateSelectionParams{ std::in_place, Terminal::SelectionDirection::Down, Terminal::SelectionExpansion::Char };
            default:
                [[fallthrough]];
            }
        }
    }
    return std::nullopt;
}
```

* `[[fallthrough]]`
* `std::pair` if you got exactly 2 items in a tuple (the most common case anyways)
* idk if you like the above more, but I prefer it (it's easier to understand IMO)
* I'd suggest making this a private member function to save yourself from writing `Terminal::` again and again.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 8]
```diff
@@ -359,28 +427,38 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
         if (HasSelection() &&
             !KeyEvent::IsModifierKey(vkey) &&
             vkey != VK_SNAPSHOT &&
             keyDown)
         {
+            // try to update the selection
+            if (const auto updateSlnParams{ ConvertKeyEventToUpdateSelectionParams(modifiers, vkey) })
+            {
+                auto lock = _terminal->LockForWriting();
+                _terminal->UpdateSelection(std::get<0>(*updateSlnParams), std::get<1>(*updateSlnParams));
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:447]: If you use pair, use first and second instead.

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 9]
```diff
@@ -151,8 +151,9 @@ void Terminal::SetSelectionEnd(const COORD viewportPos, std::optional<SelectionE
     // Otherwise, we may accidentally expand during other selection-based actions
     _multiClickSelectionMode = newExpansionMode.has_value() ? *newExpansionMode : _multiClickSelectionMode;
 
-    bool targetStart = false;
-    const auto anchors = _PivotSelection(textBufferPos, targetStart);
+    std::pair<COORD, COORD> anchors;
+    bool targetStart;
+    std::tie(anchors.first, anchors.second, targetStart) = _PivotSelection(textBufferPos);
```
Comment (lhecker) [inline src/cascadia/TerminalCore/TerminalSelection.cpp:156]: I prefer the previous code.

File: src/buffer/out/textBuffer.cpp (hunk 6 above)
Comment (carlos-zamora) [reply src/buffer/out/textBuffer.cpp:1439]: Yeeeeaaaaah. So, a11y and selection have a lot of similarities and very small/important differences. Some highlights include...
| | Selection | A11y |
|--|--|--|
| endpoint representation | both endpoints are inclusive; start <= end | inclusive start; exclusive end; start <= end |
| definition of a word | a run of legible characters | run of legible characters + run of whitespace after it (until next word starts) |

I just chose to name the parameter `accessibilityMode` to make it clear what and when we're only doing some extra work for a11y.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 10]
```diff
@@ -359,28 +427,38 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
-        // flow through to the terminal.
+        // Update the selection, if it's present
         // GH#6423 - don't dismiss selection if the key that was pressed was a
         // modifier key. We'll wait for a real keystroke to dismiss the
         // GH #7395 - don't dismiss selection when taking PrintScreen
         // selection.
-        // GH#8522, GH#3758 - Only dismiss the selection on key _down_. If we
-        // dismiss on key up, then there's chance that we'll immediately dismiss
+        // GH#8522, GH#3758 - Only modify the selection on key _down_. If we
+        // modify on key up, then there's chance that we'll immediately dismiss
```
Comment (zadjii-msft) [inline src/cascadia/TerminalControl/ControlCore.cpp:436]: \<showerthought>


If we could handle this modification on key up, then could we do something like bind `markMode` to `shift+left`, and have the `shift+left` keydown _enter mark mode_, and the keyup _extend the selection to the left_? Essentially getting quickedit-like selection?

This is of course, crazy, but something to think about

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:436]: HECK IT DOESN"T EVEN MATTER

a user could just bind markmost to shift+left, and that would start a selection at the cursor, and the subsequent keypresses would extend it without hitting the keybinding, which is actually just like conhost. Brilliant 🥂

**Review by lhecker — APPROVED** [review]

**Comment by zadjii-msft** [comment]
shipit
